### PR TITLE
Separation compilation and execution

### PR DESCRIPTION
**Pull request by partouf**
As part of the journey to improve execution support, this is my attempt at separating the steps.

I think I succeeded, but it might be a bit of a mess, still. So comments are welcome.

### CODE REVIEW

File: lib/base-compiler.js [hunk 1]
```diff
@@ -258,6 +262,65 @@ class BaseCompiler {
         return tooling;
     }
 
+    buildExecutable(compiler, options, inputFilename, execOptions) {
+        // default implementation, but should be overridden by compilers
+        return this.runCompiler(compiler, options, inputFilename, execOptions);
+    }
+
+    getOrBuildExecutable(key) {
+        // todo: ask cache for executable if it already exists
```
Comment (mattgodbolt) [inline lib/base-compiler.js]: :+1: here . This is the main reason the S3 cache exists: then a compile on one node can be executed on another. And eventually we can separate execution onto different nodes entirely (the `sandbox` call might eventually send a message to a separate node with the hash as a parameter, for example).

It shouldn't be too tricky to check the cache here I _think_, right? But of course we need to be careful to use the right things in the key, and we need to ensure the resulting cache answer is executable...so I see why you might punt on this for now :)

File: lib/base-compiler.js [hunk 2]
```diff
@@ -269,11 +332,24 @@ class BaseCompiler {
             delete filters.binary;
         }
 
-        const key = {compiler: this.compiler, source, options, backendOptions, filters, tools};
+        const executeParameters = options.executeParameters || [];
```
Comment (mattgodbolt) [inline lib/base-compiler.js]: Good thinking: the cache key doesn't need the execution parameters.

File: lib/compilers/pascal.js [hunk 3]
```diff
@@ -148,7 +152,7 @@ class FPCCompiler extends BaseCompiler {
     execBinary(executable, result, maxSize) {
         executable = path.join(path.dirname(executable), "prog");
```
Comment (mattgodbolt) [inline lib/compilers/pascal.js]: Can use `getExecutableFilename` here now :)

File: lib/base-compiler.js [hunk 4]
```diff
@@ -408,6 +484,17 @@ class BaseCompiler {
                             return result;
                         })
                         .then(result => {
+                            if (doExecute) {
```
Comment (mattgodbolt) [inline lib/base-compiler.js]: I'm not sure whether we should cache first, execute later? Or cache while we execute. Seems like we shouldn't wait to put this in the cache until execution works. Also, we probably want to cache the compilation results even if the execute fails? Again, not sure here, don't think there's a right answer, but wondered if you had any thoughts?

Comment (partouf) [reply lib/base-compiler.js]: I think the most ideal thing would be if executables would be cached as early as possible.

For example when some compilers actually need to produce executables to see the asm, or when someone clicks binary mode and it produces an executable (some only produce object files) - if we cache it then and there, you could get it from the cache without recompiling when someone hits execute.

But when execution happens I think we should probably cache the execution result as well maybe? But then we can't cleanly support things like time or rand() related executables (unless the commandline parameters are changed in between tries).

Comment (partouf) [reply lib/base-compiler.js]: Wait, I understand what you mean now, I think.
I will see and test how early I can call it...

Comment (partouf) [reply lib/base-compiler.js]: Changed some things around, but kept the waiting for execution result. Need to think about it some more

Comment (partouf) [reply lib/base-compiler.js]: removed the execResult from the caching

**Pull request merged**